BOSCH144: Add IRac class support

NicoThien opened the pull request:
BOSCH144: Add IRac class support
Power, Mode, Temp, Fan, Quiet is working
I want to add: SwingV, Turbo, Clean, Beep/LED (all separate 96-bit messages)

NicoThien committed:
BOSCH144: Add IRac class support

BOSCH144: Add IRac class support
Power, Mode, Temp, Fan, Quiet is working
I want to add: SwingV, Turbo, Clean, Beep/LED (all separate 96-bit messages)

NicoThien commented:
Bosch CL3000i-Set 26 E Remote: RG10A(G2S)BGEF
https://github.com/crankyoldgit/IRremoteESP8266/issues/1787#issuecomment-1167221417

NicoThien committed:
Update IRac.cpp

correction of a copy/paste error

crankyoldgit commented on src/IRac.cpp:
```diff
@@ -459,6 +463,52 @@ void IRac::argo(IRArgoAC *ac,
+/// @param[in] swingv The vertical swing setting.                                  *** todo
+/// @param[in] quiet Run the device in quiet/silent mode.
+/// @param[in] turbo Run the device in turbo/powerful mode.                        *** todo
+/// @param[in] clean Turn on the self-cleaning mode. e.g. Mould, dry filters etc   *** todo
```
Don't include the not-yet supported (i.e. todo) parameters. It will confuse people & the documentation system.
We just add them as they get supported.

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+  if( nbytes % kBosch144BytesPerSection != 0 )return;
   // Set IR carrier frequency
-  enableIROut(kBoschFreq);
+  enableIROut( kBoschFreq );
```
Please match the coding layout style of the rest of the code.
The `cpplint`er may reject it also.
e.g.
```c++
  if (nbytes % kBosch144NrOfSections != 0) return;
  enableIROut(kBoschFreq);
```
rather than
```c++
  if( nbytes % kBosch144BytesPerSection != 0 )return;
  enableIROut( kBoschFreq );
```

This goes generally for all of the changes.

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+/// @return A valid code for this protocol based on the current internal state.
+uint8_t* IRBosch144AC::getRaw( void ) 
+{ 
+  return _.raw; 
```
You should do the inverting/checksum calculations before returning the pointer.

NicoThien replied:
I added the functions to getRaw()

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+
+/// Set the desired operation mode.
+/// @param[in] mode The desired operation mode.
+void IRBosch144AC::setMode( const uint8_t mode )
```
Where is `getMode()`?

NicoThien replied:
getMode() is now also available

crankyoldgit commented on src/ir_Bosch.h:
```diff
@@ -0,0 +1,166 @@
+// Copyright 2022 David Conran
```
Your name here, not mine. It's your work & effort! ;-)

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+  _.ChecksumS3 += _.raw[ 13 ];
+  _.ChecksumS3 += _.raw[ 14 ];
+  _.ChecksumS3 += _.raw[ 15 ];
+  _.ChecksumS3 += _.raw[ 16 ];
```
You can use `sumBytes()` from:
https://github.com/crankyoldgit/IRremoteESP8266/blob/2bad02785274a54b5208228d80c159e0826f193a/src/IRutils.h#L34-L35
e.g.
```c++
  _.ChecksumS3 =  sumBytes(_.raw + 12, 5);
```
or
```c++
  _.ChecksumS3 =  sumBytes(&(_.raw[12]), 5);
```

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+  _.InnvertS1_3 = ~_.raw[ 4 ];
+  _.InnvertS2_1 = ~_.raw[ 6 ];
+  _.InnvertS2_2 = ~_.raw[ 8 ];
+  _.InnvertS2_3 = ~_.raw[ 10 ];
```
As they are fixed/entire bytes, why not use a `for` loop.
e.g.
```c++
  for (uint16_t i = 0; i <= 10; i += 2) {
    _raw[i + 1] = ~_.raw[i];
  }
```

NicoThien replied:
replaced

crankyoldgit commented on src/ir_Bosch.h:
```diff
@@ -0,0 +1,166 @@
+  void setFan(const uint16_t speed);
+  uint16_t getFan(void) const;
+  void setMode(const uint8_t mode);
+  uint8_t getMode(void) const;
```
Missing?

NicoThien replied:
getMode() and getFan() is now also available

crankyoldgit commented on src/ir_Bosch.h:
```diff
@@ -0,0 +1,166 @@
+  stdAc::state_t toCommon(void) const;
+  String toString(void) const;
+  void setInvertBytes();
+  void setCheckSumS3();
```
Do these need to be public? Especially if they are added to `getRaw()`

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+void IRBosch144AC::setFan( const uint16_t speed )
+{
+  _.FanS1 = speed >> 6;         // save 3 bits in Section1
+  _.FanS2 = speed >> 6;         // Section2 = Section1
```
`  _.FanS1 = _.FanS2 = speed >> 6;`

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+/// @param[in] speed The desired setting.
+void IRBosch144AC::setFan( const uint16_t speed )
+{
+  _.FanS1 = speed >> 6;         // save 3 bits in Section1
```
As speed is a `uint16_t`, you're saving more than just 3 in to Section 1. Either your comment is wrong, or your code is.

crankyoldgit replied:
Ignore my comment. I noticed it's only 3 bits in the protocol description/structure.

crankyoldgit commented on src/ir_Bosch.h:
```diff
@@ -0,0 +1,166 @@
+    0b10010,  // 27C
+    0b10000,  // 28C
+    0b10100,  // 29C
+    0b10110   // 30C
```
Wow. That's just a bizarre sequence. Nothing to fix here. Just a comment.

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+
+bool IRBosch144AC::getPower( void ) const 
+{
+  return true;
```
Shouldn't you check if the `_.raw`/state code isn't the Off message and if `powerFlag` is `True`?

NicoThien replied:
I guess I didn't realize then that I wasn't done yet.

crankyoldgit replied:
Well, this protocol seems more complicated than most. So, kind of expected.

NicoThien replied:
getPower() now queries the state of powerFlag.

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+/// @param[in] new_code A valid code for this protocol.
+void IRBosch144AC::setRaw( const uint8_t new_code[] ) 
+{
+  std::memcpy( _.raw, new_code, kBosch144StateLength );
```
You probably also want:
```c++
  std::memcpy( _.raw, new_code, kBosch144StateLength);
  powerFlag = getPower();
```

NicoThien replied:
getPower() only queries the value of powerFlag

powerFlag is set in IRac.cpp line 481

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,298 @@ const uint16_t kBoschZeroSpace = 571;
+/// Reset the internal state to a fixed known good state.
+void IRBosch144AC::stateReset( void ) 
+{
+  setRaw( kBosch144DefaultState );
```
Also set an intial value for `powerFlag` too.

NicoThien replied:
is now set to "true" there.

crankyoldgit reviewed (changes requested):
Top effort adding the detailed protocol decode. That's an excellent start!

NicoThien commented:
I will try to fix the errors.

NicoThien committed:
Update IRac.cpp

Edit coding layout style

NicoThien committed:
Update ir_Bosch.h

Edit coding layout style

NicoThien committed:
Update ir_Bosch.h

Edit coding layout style

NicoThien committed:
Update and rename src/ir_Bosch.h to Quelle/ir_Bosch.h

Edit coding layout style

NicoThien committed:
Rename Quelle/ir_Bosch.h to src/ir_Bosch.h

NicoThien committed:
Update ir_Bosch.cpp

Edit coding layout style

NicoThien committed:
Update ir_Bosch.cpp

Edit coding layout style

NicoThien committed:
Update ir_Bosch.cpp

Edit coding layout style

NicoThien committed:
Update ir_Bosch.cpp

Edit coding layout style

NicoThien committed:
Update ir_Bosch.cpp

NicoThien committed:
Update ir_Bosch.cpp

add getMode() and grtFan()

NicoThien committed:
Update ir_Bosch.cpp

crankyoldgit commented:
FYI, you'll also need to update the unit tests for this protocol, now that it has detailed support mostly.
eg.
https://github.com/crankyoldgit/IRremoteESP8266/runs/7417257804?check_suite_focus=true#step:6:2292

NicoThien committed:
Update ir_Bosch.cpp

powerFlag is now set to "true" in the "state reset".

NicoThien committed:
Update ir_Bosch_test.cpp

NicoThien committed:
Update ir_Bosch_test.cpp

NicoThien committed:
Update ir_Bosch_test.cpp

NicoThien committed:
Update ir_Bosch_test.cpp

NicoThien committed:
Edit Test

NicoThien committed:
Add files via upload

NicoThien committed:
Add files via upload

NicoThien committed:
Add files via upload

NicoThien committed:
Update ir_Bosch.cpp

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -8,9 +8,11 @@
 //   Brand: Bosch,  Model: RG10A(G2S)BGEF remote
```
Now that you've got a ir_Bosch.h file, and the "Supports" data is listed in that, you will need to remove the "Supports" comment section from the .cpp file. Otherwise it will fail one of our automated checks.
i.e. "Supports" info always goes into the .h file if it exists, if not, then into the .cpp file.

crankyoldgit replied:
Also, please add yourself to the Copyright section. You've added significant work too. ;-)

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -8,9 +8,11 @@
-#endif
+#endif*/
+#include <cstring>
+#include "ir_Bosch.h"
```
Why are these changes needed?

NicoThien replied:
Without `<cstring>` I get `std::memcpy(_.raw, new_code, kBosch144StateLength);` the error message "std has no member memcpy". I thought `<cstring>` would replace `<string>`.
My programming skills are very limited.
My English too.

crankyoldgit replied:
`<cstring>` is fine, e.g.:
https://github.com/crankyoldgit/IRremoteESP8266/blob/85d587fbbf804c25c15632849386091e712a6cae/src/ir_Samsung.cpp#L14-L17

it was the other bit I was refering to. i.e. the removal of the `#ifndef` etc.

NicoThien replied:
is back in

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
-  if (nbytes % kBosch144NrOfSections != 0) return;
-
+  // nbytes is required to be a multiple of kBosch144BytesPerSection.
+  if (nbytes % kBosch144BytesPerSection != 0)return;
```
A space before the `return;` please.

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
+/// @param[in] new_code A valid code for this protocol.
+void IRBosch144AC::setRaw(const uint8_t new_code[]) {
+  std::memcpy(_.raw, new_code, kBosch144StateLength);
+}
```
What happens if someone sets the "Off" message as the raw state?
You need to handle that length, and to set the appropriate `powerFlag` value.

NicoThien replied:
Could this work?
```suggestion
void IRBosch144AC::setRaw(const uint8_t new_code[]) {
  if (sizeof(new_code) == kBosch144StateLength) {
    setPower(true);
    std::memcpy(_.raw, new_code, kBosch144StateLength);
  } else {
    if (new_code == kBosch144Off) setPower(false);
  }
}
```

NicoThien replied:
it doesn't work

NicoThien replied:
`void IRBosch144AC::setRaw(const uint8_t new_code[]) {`
`  if (new_code == kBosch144Off) {`
`    setPower(false);`
`  } else {`
`    std::memcpy(_.raw, new_code, kBosch144StateLength);`
`  }`
`}`
should work

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
+
+uint8_t IRBosch144AC::getTemp(void) const {
+  uint8_t temp = (_.TempS1 << 1) + _.TempS3;
+  uint8_t retemp = 0;
```
Is a temp of `0` a good default? e.g. what happens if it can't find the temp value in the lookup table?

NicoThien replied:
changed to 25

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
+
+uint16_t IRBosch144AC::getFan(void) const {
+  uint16_t speed = (_.FanS1 << 6) + _.FanS3;
+  return speed;
```
just:
```
  return (_.FanS1 << 6) + _.FanS3;
```

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
+
+uint8_t IRBosch144AC::getMode(void) const {
+  uint8_t mode = (_.ModeS1 << 1) + _.ModeS3;
+  return mode;
```
`  return (_.ModeS1 << 1) + _.ModeS3;`

NicoThien replied:
changed

crankyoldgit commented on src/ir_Bosch.h:
```diff
@@ -0,0 +1,173 @@
+/*const uint8_t kBosch144DefaultState[kBosch144StateLength] = {
+  0xB2, 0x00, 0x00, 0x00, 0x00, 0x00,
+  0xB2, 0x00, 0x00, 0x00, 0x00, 0x00,
+  0xD5, 0x00, 0x00, 0x00, 0x00, 0x00};*/
```
Do we need this comment?

NicoThien replied:
no

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -25,37 +27,276 @@ const uint16_t kBoschZeroSpace = 571;
+                           static_cast<int>(stdAc::fanspeed_t::kMin),
+                           static_cast<int>(stdAc::fanspeed_t::kAuto),
+                           static_cast<int>(stdAc::fanspeed_t::kAuto),
+                           static_cast<int>(stdAc::fanspeed_t::kMedium));
```
Rather than use the "common" fan speeds. You're better off using the `kBosch144Fan20` etc values you get directly from `getFan()`.
This way a user is told what value to pass to `setFan()` in the `toString()` output. It also helps with debugging if we get some other "raw" fan speed value.

NicoThien replied:
addFanToString() expects 8-bit arguments
kBosch144Fan20 has 9 bits

crankyoldgit replied:
Good point!

crankyoldgit reviewed (changes requested):
Looking much better. Nearly there!

NicoThien committed:
Update ir_Bosch.h

- commented out lines removed
- change  setInvertBytes() & setCheckSumS3() to private:

NicoThien committed:
Update ir_Bosch.cpp

- add copyright
- simplify getMode() & getFan()
- initial temperature in getTemp() 25

NicoThien committed:
Update ir_Bosch.cpp

removed
// Supports:
//   Brand: Bosch,  Model: CL3000i-Set 26 E A/C
//   Brand: Bosch,  Model: RG10A(G2S)BGEF remote

crankyoldgit commented:
Friendly ping. How it going?

NicoThien committed:
Update src/ir_Bosch.cpp

NicoThien committed:
Update ir_Bosch.cpp

NicoThien commented:
If there are no other change requests...
It works for me.

crankyoldgit commented on src/ir_Bosch.cpp:
```diff
@@ -99,11 +99,10 @@ uint8_t* IRBosch144AC::getRaw(void) {
-  if (sizeof(new_code) == kBosch144StateLength) {
-    setPower(true);
-    std::memcpy(_.raw, new_code, kBosch144StateLength);
+  if (new_code == kBosch144Off) {
```
This doesn't do what you think it does.
It's comparing if the two pointers are the same, not the contents of the two arrays (where the pointers point to) are the same. i.e. You're checking if the two arrays have the same memory address. They won't.
You need to use [`std::memcmp()`](https://en.cppreference.com/w/cpp/string/byte/memcmp) to compare the contents.

You will also need to pass in the `length` of the array as a parameter to `setRaw()` and make sure you don't exceed the size of `_.raw`

Something like:
```c++
/// Set the internal state from a valid code for this protocol.
/// @param[in] new_code A valid code for this protocol.
/// @param[in] length Size of the array being passed in in bytes.
void IRBosch144AC::setRaw(const uint8_t new_code[], const uint16_t length) {
  const uint16_t len = min(length, kBosch144StateLength);
  if (std::memcmp(kBosch144Off, new_code, min(sizeof(kBosch144Off), len))) == 0) // Is it an off message?
    setPower(false);  // It is.
  else
    setPower(true);
  std::memcpy(_.raw, new_code, len);
}
```
and in `ir_Bosch.h`:
```c++
  void setRaw(const uint8_t new_code[], const uint16_t length = kBosch144StateLength);
```

NicoThien committed:
Add files via upload

NicoThien committed:
Update ir_Bosch.cpp

NicoThien committed:
Update ir_Bosch.h

crankyoldgit commented:
FYI, the changes mentioned above have now been included in the new [v2.8.3 release](https://github.com/crankyoldgit/IRremoteESP8266/releases/tag/v2.8.3) of the library.
